src/dynamixel_hardware_interface.cpp

```diff
@@ -114,26 +125,50 @@ namespace dynamixel
     **/
     void DynamixelHardwareInterface::read_joints()
     {
-        try
-        {
-            for (unsigned i=0; i<_dynamixel_ids.size(); i++)
-            {
-                dynamixel::Status status;
+        for (unsigned i = 0; i < _dynamixel_servos.size(); i++) {
+            std::map<long long int, std::string>::iterator dynamixel_iterator = _dynamixel_map.find(_dynamixel_servos[i]->id());
+            if (dynamixel_iterator == _dynamixel_map.end())
+                continue;
+            dynamixel::StatusPacket<dynamixel::protocols::Protocol1> status;
+            try {
                 // current position
-                _dynamixel_controller.send(dynamixel::ax12::GetPosition(_dynamixel_ids[i]));
-                _dynamixel_controller.recv(_read_timeout, status);
-                _joint_angles[i] = status.decode16();
-
-                // current speed
-                _dynamixel_controller.send(dynamixel::ax12::GetSpeed(_dynamixel_ids[i]));
-                _dynamixel_controller.recv(_read_timeout, status);
-                _joint_velocities[i] = status.decode16();
+                _dynamixel_controller.send(_dynamixel_servos[i]->get_present_position_angle());
+                _dynamixel_controller.recv(status);
             }
-        }
-        catch (Error& e)
-        {
-            ROS_ERROR_STREAM("Caught a Dynamixel exception while getting the actuator's state:\n"
-                << e.msg());
+            catch (dynamixel::errors::Error& e) {
+                ROS_ERROR_STREAM("Caught a Dynamixel exception while getting  " << _dynamixel_map[_dynamixel_servos[i]->id()] << "'s position\n" << e.msg());
+            }
+            if (status.valid()) {
+                _joint_angles[i] = _dynamixel_servos[i]->parse_present_position_angle(status);
+
+                std::map<long long int, double>::iterator dynamixel_corrections_iterator = _dynamixel_corrections.find(_dynamixel_servos[i]->id());
+                if (dynamixel_corrections_iterator != _dynamixel_corrections.end()) {
+                    _joint_angles[i] -= dynamixel_corrections_iterator->second;
+                }
+            }
+            else {
+                ROS_WARN_STREAM("Did not receive any data when reading " << _dynamixel_map[_dynamixel_servos[i]->id()] << "'s position");
+            }
+
+            // TO-DO: Make it work in V2 libdynamixel
```

Comment:
It is possible, in 3 steps:
- Find for each dynamixel model what is the unit of the current speed field (they may differ for model to model). The issue with this is that not every model is well documented, for example in the Pros I couldn't find it, but maybe with the Dynamixel Software we can get it.
- Add a double field called something like `current_speed_unit` in the traits of all models, putting the specific value for each one of them.
- Add the functions `get_present_speed_m_s` and `parse_present_speed_m_s` both in `BaseServo` as virtual functions that throw exceptions (like in `get_present_position_angle`), and in `Servo` as the implementation and also as static functions available for the models